enhancements/cluster-api/crd-compatibility-checker.md

```diff
@@ -0,0 +1,655 @@
+  This indicates that the CRD exists and is compatible with this requirement, but `Message` contains one or more warning messages.
+  It is associated with a status of `True`.
+  * `Compatible` -
+  This indicates that the CRD exists and is compatible with this requirement.
+  It is associated with a status of `True`.
+
+#### Validating webhooks
+
+##### ClusterResourceDefinition
+
+Create, Update, and Delete operations on CRDs will either be denied or emit a warning depending on the setting of `crdSchemaValidation.action`.
+
+Create and Update will deny/warn if the created or updated CRD would not be compatible with any CRDCompatibilityRequirement which references it in `status.crd`.
+
+Delete will deny/warn if any CRDCompatibilityRequirement references it in `status.crd`.
+
+##### Dynamically created by `objectSchemaValidation`
+
+[!NOTE]
+Implementation of `objectSchemaValidation` will be deferred to a second phase to be implemented after the CRD validation webhook is fully implemented and integrated.
+
+If a CRDCompatibilityRequirement defines `objectSchemaValidation` we will dynamically create a new `ValidatingWebhookConfiguration` to implement it.
+This webhook will ensure that matching objects conform to the CRD given in `compatibilityCRD`, which may be different to the current CRD version.
+It will either warn or deny admission based on the setting of `objectSchemaValidation.action`.
+
+This webhook will have the same lifecycle as the CRDCompatibilityRequirement which created it.
+
+The webhook will only be called for objects matching the selectors in `objectSchemaValidation`.
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+The Hypershift management cluster is one of the primary intended users of this feature.
+
+The feature should work correctly in a managed cluster, although we don't expect CAPI CRDs to be managed there so it wouldn't be an initial target.
+
+#### Standalone Clusters
+
+This enhancement is relevant for standalone clusters as it addresses scenarios where external workloads manage CRDs used by OpenShift payload components.
+
+#### Single-node Deployments or MicroShift
+
+As for managed clusters, the feature should work correctly on SNO but we would not initially target those deployments.
+
+### Implementation Details/Notes/Constraints
+
+N/A
+
+### Risks and Mitigations
+
+- **Risk**:
+  Webhook failures could block CRD or object operations
+
+  **Mitigation**:
+  Run the typical deployment configuration with 2 members which coordinate via leader election.
+  Ensure that the webhooks can safely run HA in both members.
+
+- **Risk**:
+  Performance impact from additional admission webhook calls
+  
+  **Mitigation**:
+  CRD updates are infrequent enough that this is unlikely to be a concern.
+  For object schema validation there is potential for impact during certain phases of cluster activity.
+  We will optimize performance beyond the initial implementation only if it proves necessary.
```

Comment:
Have we considered implementing any metrics to track the impact the webhooks are going to have?

Reply:
I expect that metric does exist yes, but I don't know if that will proportion correctly where the time is being spent